docs/develop/compiler.rst

```diff
@@ -0,0 +1,417 @@
+                compile_scope_inline_asm(comp, s, MP_PASS_EMIT);
+            }
+        } else
+
+The inline assembler code comprises assembly instructions in a Python function.
+See `the inline assembler tutorial 
+<https://docs.micropython.org/en/latest/pyboard/tutorial/assembler.html#pyboard-tutorial-assembler>`_ 
+for more details.
+
+Fourth, fifth and sixth passes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The other two passes compute the stack and code size, while the last pass emits the final code:
+
+.. code-block:: c
+   
+   compile_scope(comp, s, MP_PASS_STACK_SIZE);
+
+   if (comp->compile_error == MP_OBJ_NULL) {
+        compile_scope(comp, s, MP_PASS_CODE_SIZE);
+   }
+
+   if (comp->compile_error == MP_OBJ_NULL) {
+    compile_scope(comp, s, MP_PASS_EMIT);
+   }
+
+Before these passes, there is a selection for the type of code to be emitted which can either be native or
+bytecode.
+
+.. code-block:: c
+
+   switch (s->emit_options) {
+
+        #if MICROPY_EMIT_NATIVE
+        case MP_EMIT_OPT_NATIVE_PYTHON:
+        case MP_EMIT_OPT_VIPER:
+            if (emit_native == NULL) {
+                emit_native = NATIVE_EMITTER(new)(&comp->compile_error, &comp->next_label, max_num_labels);
+            }
+            comp->emit_method_table = NATIVE_EMITTER_TABLE;
+            comp->emit = emit_native;
+            break;
+        #endif // MICROPY_EMIT_NATIVE
+
+        default:
+            comp->emit = emit_bc;
+            #if MICROPY_EMIT_NATIVE
+                comp->emit_method_table = &emit_bc_method_table;
+            #endif
+            break;
+   }
+
+The bytecode option is the default but something unique to note for the native code option is that there is another
+option via ``VIPER``. See the :ref:`Emitting native code <emitting_native_code>` section for
+more details on viper annotations.
+
+Emitting bytecode
+-----------------
+
+For every statement in the code, there is a corresponding function to emit MicroPython bytecode. 
+This function should be written in ``py/emitbc.c``. The implementation of this function looks similar
+to this:
+
+.. code-block:: c
+   
+   void mp_emit_bc_yield(emit_t *emit, int kind) {
+        MP_STATIC_ASSERT(MP_BC_YIELD_VALUE + 1 == MP_BC_YIELD_FROM);
+        emit_write_bytecode_byte(emit, -kind, MP_BC_YIELD_VALUE + kind);
+        emit->scope->scope_flags |= MP_SCOPE_FLAG_GENERATOR;
+   }
+
+We use the ``yield`` statement for an example here but the implementation details are similar for other statements.
+The method ``emit_write_bytecode_byte()`` is a wrapper around the main function ``emit_get_cur_to_write_bytecode()``
+that all functions must call to emit byte code.
+
+.. _emitting_native_code:
+
+Emitting native code
+---------------------
+
+Similar to how bytecode is generated, there should be a corresponding function in ``py/emitnative.c`` for each
+code statement:
+
+.. code-block:: c
+
+   STATIC void emit_native_yield(emit_t *emit, int kind) {
```

Comment:
this is a pretty complicated function, would it be better to use as an example a simpler function like unary op?  (so use `mp_emit_bc_unary_op` in the bytecode section above, and `emit_native_unary_op` here?)